src/UglyToad.PdfPig/Writer/PdfStreamWriter.cs

```diff
@@ -0,0 +1,166 @@
+                var referenceToken = pair.Key;
+                var token = pair.Value;
+                var offset = Stream.Position;
+                var obj = new ObjectToken(offset, referenceToken.Data, token);
+
+                TokenWriter.WriteToken(obj, Stream);
+
+                offsets.Add(referenceToken.Data, offset);
+
+                if (catalogToken == null && referenceToken == catalogReference)
+                {
+                    catalogToken = obj;
+                }
+            }
+
+            if (catalogToken == null)
+            {
+                throw new Exception("Catalog object wasn't found");
+            }
+
+            // TODO: Support document information
+            TokenWriter.WriteCrossReferenceTable(offsets, catalogToken, Stream, null);
+        }
+
+        public IndirectReferenceToken WriteToken(IToken token, int? reservedNumber = null)
+        {
+            // if you can't consider deduplicating the token. 
+            // It must be because it's referenced by his child element, so you must have reserved a number before hand
+            // Example /Pages Obj
+            var canBeDuplicated = !reservedNumber.HasValue;
+            if (!canBeDuplicated)
+            {
+                if (!reservedNumbers.Remove(reservedNumber.Value))
+                {
+                    throw new InvalidOperationException("You can't reuse a reserved number");
+                }
+
+                // When we end up writing this token, all of his child would already have been added and checked for duplicate
+                return AddToken(token, reservedNumber.Value);
+            }
+
+            var reference = FindToken(token);
+            if (reference == null)
+            {
+                return AddToken(token, CurrentNumber++);
+            }
+
+            return reference;
+        }
+        
+        public int ReserveNumber()
+        {
+            var reserved = CurrentNumber;
+            reservedNumbers.Add(reserved);
+            CurrentNumber++;
+            return reserved;
+        }
+
+        public IndirectReferenceToken ReserveNumberToken()
+        { 
+            return new IndirectReferenceToken(new IndirectReference(ReserveNumber(), 0));
+        }
+        
+        public byte[] ToArray()
+        {
+            if (!Stream.CanSeek)
+                throw new NotSupportedException($"{Stream.GetType()} can't seek");
+            
+            var currentPosition = Stream.Position;
+            Stream.Seek(0, SeekOrigin.Begin);
+
+            var bytes = new byte[Stream.Length];
+
+            // Should we slice the reading into smaller chunks?
+            if (Stream.Read(bytes, 0, bytes.Length) != bytes.Length)
+                throw new Exception("Unable to read all the bytes from stream");
+
+            Stream.Seek(currentPosition, SeekOrigin.Begin);
+
+            return bytes;
+        }
+
+        public void Dispose()
+        {
+            Stream?.Dispose();
```

Comment:
I generally like to have a flag for this, if the user has provided the stream we probably shouldn't dispose it ourselves. For example they might be writing a streaming web response or streaming some other form of data and they want to continue to use the stream, for example to write an end-of-data marker after using it for our method. On the other hand if we've instantiated our own memory stream we're allowed to dispose it. But since memorystream doesn't require disposing we can probably remove the dispose entirely, or just null our reference.

Reply:
Done deal